### ISSUE TITLE
Remote does not work

### ISSUE THREAD

**Issue by mittorn** [lunixbochs/glshim#94]
After adding some initialisation to glx.c
```
	state.remote=1;
	remote_spawn("/home/mittorn/glshim/bin/libgl_remote");
```
i found that command buffer works wrong.
libgl_remote terminates with "Invalit remote command" error.
I found that command length always equals 2.
Fut after changing condition "if (c.args != 3)" rendering does not work.
It simply "reads" these two commands in infinite cycle after some moment:
```
185 glColor3ub
2066 glTextureStorage2DMultisampleEXT
185 glColor3ub
```
Last executed command are:
```
1573 glProgramUniform1ui64NV
86 glBinormal3fEXT
1925 glTexCoord2f
1566 glProgramUniform1i
1188 glMatrixLoadTransposefEXT
657 glGetColorTable
```
Maybe i missed something?

**Comment by lunixbochs** [comment]
You manually grabbed an unfinished feature branch and it didn't work? :)

It's not done yet, and I'm pretty sure there's unpushed code. Do you want to help develop it?

**Comment by lunixbochs** [comment]
I'm really curious what you're running this on.

Also, even though it can send some draw calls to the remote process, glx is not proxied yet, so there's no way you'll get a proper GL context.

**Comment by mittorn** [comment]
I know this. I want to use proxy on android. It might be useful with gnu/linux chroot on android.
I can create context on remote side (Android app) and use different shared memory technology when shm unavailiable, but it is sad that proxy does not work.

**Comment by mittorn** [comment]
As i understand, it must proxy gl calls on newer exagear versions as replace of old exagear proxy, because mobile exagear's translator cannot be run is same process with GLES driver and dalvik vm for some reason. Old proxy used special syscall as very effective solution without memory copy. Am i right?

**Comment by lunixbochs** [comment]
Did you actually have the old exagear GL proxy? I didn't see anyone but me with that build. It did use a syscall, but broke on most GL drivers due to the very memory unsafety that made it so fast.

I worked on this quite a bit last night. I won't be surprised if I have a demo ready this weekend. proxy2 doesn't actually use any shm() syscalls. It mmap()s a shared file into both processes and uses it as a ring buffer for commands.

**Comment by mittorn** [comment]
I lost old repo with broken mmc card, but i saw proxy client code.
I don't know how to use it with any version of ExaGear (Desktop or Mobile) that i can run.
Both desktop and mobile has translator in static binary that cannot load any GLES drivers.
As supported GL version (GL1.x) is used in games with low graphics requirements, it is possible to use shm or even socket pipe to transport data.
For example, Half-Llife 1 works prefectly with AIGLX and slow wifi connection, when used arm cortex-a8 device with exagear and X-server on remote Intel PC.

**Comment by lunixbochs** [comment]
It should be in a pretty reasonable state now. Many of my tests work now. Can you test and report back?

**Comment by mittorn** [comment]
On https://github.com/lunixbochs/glshim/commit/26be7eb62167406415a1206deee659bcafe36ab7 only basic drawing. Texture loading does not work correctly. All textures are broken. glxgears may remote crash with invalid command after few seconds, it seems to break arguments randomly. But it is better then was at last try. With fps limit it seems to be more stable (less frames, less crashes). Local process does not crash at all.

**Comment by mittorn** [comment]
Yesterday only 2d menu displayed with noise and glxgears not drawed at all.

**Comment by lunixbochs** [comment]
I'm working on semaphore integration which should fix the remaining sporadic crashes (there's currently a race condition that will crash the remote end).

Some GL calls are likely to outright crash, as I don't handle every function that passes pointer arguments yet.

Fixing texture issues is the next thing on my list. I also noticed the lighting in glxgears seems weird to me. Might be matrix related.

**Comment by mittorn** [comment]
I have some lighting problems in glxgears too.
After update remote seems does not crash, but hangs sometimes. It seems limited to 30fps.

**Comment by lunixbochs** [comment]
30fps is caused by bad vsync interaction with 5d896400cb0ae944651969d66d2de869ac15fe04 - it will be fixed by #103.

The hangs will be fixed by my WIP branch, where I rewrote the RPC layer from scratch.

**Comment by lunixbochs** [comment]
Both of these are fixed and merged.

**Comment by lunixbochs** [comment]
Cannot reproduce the original issue. Closing. ;)

(Just merged proxy2 into unstable at cc12f34e45a819151043eede08ec110bbd311967.)
